### ISSUE TITLE
lnd-0.14.0-rc1 hangs after a few failed rebalance attempts

### ISSUE THREAD

**Issue by xmrk-btc** [lightningnetwork/lnd#5914]
### Background

Some subsystems of lnd stop working after I run rebalance-lnd and a few rebalance attempts fail (rebalance lnd reports "Temporary payment failure"). In particular, CRTR subsystem stops writing any logs even in debug, and lnd stops responding to lncli - for example `lncli getinfo` waits indefinitely. And the disc activity of lnd drops significantly, from at least 200 kb/s to some 10 kb/s. But it does not hang completely, it still produces log lines of PEER, DISC, HSWC and SRVR.
The last log line produced by CRTR is
> CRTR: Searching for path to 033878501f9a4ce97dba9a6bba4e540eca46cb129a322eb98ea1749ed18ab67735, sending 300000000 mSAT

Tested twice so far, once just few minutes after lnd started and connected to peers, lnd hanged after first rebalance failure. Second time, lnd ran fine for about 10 hours while no rebalancing was done, then I started rebalance-lnd and it hanged after 3 attempts. 

### Your environment

* version of `lnd`: 0.14.0-beta.rc1
 * db.no-graph-cache NOT set
* which operating system (`uname -a` on *Nix): `Linux ubuntu 5.11.0-1021-raspi #22-Ubuntu SMP PREEMPT Wed Oct 6 17:30:38 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux`
* version of `btcd`, `bitcoind`, or other backend: bitcoin-0.21.1
* version of rebalance-lnd: commit aca3ed199cffa6901b3cd35c8d283e24f68cf259 of 31/oct/2021 it uses grpcio 1.41.1
* libc6:arm64     2.33-0ubuntu5
* Ubuntu 21.04 64-bit, last updated 31/10/2021 
* Raspberry Pi 4 with 4 GB RAM

### Steps to reproduce

1. run lnd and wait until it initializes (reads db, connect to most of peers)
2. run rebalance-lnd (https://github.com/C-Otto/rebalance-lnd) 

### Expected behaviour

lnd should keep working, forwarding, responding to lncli getinfo.

### Actual behaviour

lnd hangs, at least CRTR subsystem.

**Comment by guggero** [comment]
Thank you for the report. Would it be possible to [pull a CPU and goroutine profile](https://github.com/lightningnetwork/lnd/blob/master/docs/debugging_lnd.md) while it's in the locked-up state?

**Comment by xmrk-btc** [comment]
I am trying now to get the profile,  but cannot get lnd to the locked state, not even after 25 failed rebalance attempts. Perhaps it depends on the exact channels being rebalanced.

**Comment by xmrk-btc** [comment]
[lnd_hangs_stacktrace.txt](https://github.com/lightningnetwork/lnd/files/7453959/lnd_hangs_stacktrace.txt)
Reproduced again, attaching profile. 

I was rebalancing by sending btc from 753999994419609600 to  753772395537039360 , so perhaps there is something special about those 2 channels .

**Comment by xmrk-btc** [comment]
I generated core file, attaching gdb output where values of variables can be seen, but I do not understand it, no experience with go. Not even sure whether gdb decodes it correctly, because of the error with libthreaddb. And I could not use dlv, it complained about too new version of go ( > 1.15). 
[gdb_log.txt](https://github.com/lightningnetwork/lnd/files/7456032/gdb_log.txt)

Still, if you consider it useful, I could run some gdb commands on my core file. I know it would be easiest to just send the core file, but it will contain sensitive info (my private keys), so I prefer no go that route.

**Comment by xmrk-btc** [comment]
@bhandras Applied your patch and it seems to work, more than 100 rebalance attempts between the same pair of channels without crash (testing with `db.no-graph-cache=1`). Thank you.

Does it make sense to test with `db.no-graph-cache=0`?

**Comment by bhandras** [comment]
Hi @xmrk-btc! Thank you for taking the time to test the upcoming RC. :1st_place_medal:  

Yes it makes sense to test both with the in-memory graph cache and with the DB fallback. I found the reason why it deadlocked on your end, I'll put another commit to the draft PR to fix that too.

**Comment by Roasbeef** [comment]
> For what it's worth, when I tested with a recent master commit from yesterday, my node stopped working 

Are you also running with the graph cache _off_? We're trying to track down this other deadlock situation, and are trying to pinpoint the core issue.

**Comment by Roasbeef** [comment]
@xmrk-btc 
> Reproduced again, attaching profile.

Is this from the deadlock or right before (?) the crash?

**Comment by C-Otto** [comment]
> Are you also running with the graph cache _off_? We're trying to track down this other deadlock situation, and are trying to pinpoint the core issue.

No, I'm using the default (with cache, much faster).

**Comment by xmrk-btc** [comment]
> @xmrk-btc
> 
> > Reproduced again, attaching profile.
> 
> Is this from the deadlock or right before (?) the crash?

from the deadlock, running with graph cache on

**Comment by Crypt-iQ** [comment]
@C-Otto what hardware & OS are you running?

**Comment by C-Otto** [comment]
It's a cloud VM with Intel Xeon 64bit, 4 virtual cores, 16 GByte RAM, fast SSDs.

**Comment by bhandras** [comment]
Could you please create a [stack trace](https://github.com/lightningnetwork/lnd/blob/master/docs/debugging_lnd.md#capturing-pprof-data-with-lnd) the next time you see this @C-Otto ?

**Comment by C-Otto** [comment]
Sorry, I won't update my (prod) node before this is fixed (and I'm aware that this is a catch 22).

**Comment by C-Otto** [comment]
I decided to update to 0.14.0-beta-rc2 and, so far, haven't been able to reproduce this issue.

**Comment by xmrk-btc** [comment]
Reproduced the original problem with 0.14.0-rc2 - lnd locked up, no disk activity, lncli getinfo hangs for > 10 minutes.

Attaching profile.

Perhaps it depends on concrete channels which are being rebalanced, this is the last rebalance attempt by rebalance-lnd:
```
Trying route #13 (fee 25,800 mSAT, 128ppm)
753029125731778561 to ln2me.com / LightningTo.Me        (free, we usually charge 25ppm)
620091572788461568 to thepkbadger                       (fee    3,600 mSAT,    17ppm)
679136446747115521 to bfx-lnd0                          (fee    1,200 mSAT,     5ppm)
753772395537039360 to xmrk                              (fee   21,000 mSAT,   105ppm)
Unknown next peer
Ignoring 620091572788461568 (ln2me.com / LightningTo.Me to thepkbadger)

```

But the `ignoring 620091572788461568` message indicates that this attempt is already finished, it is looking for a new route to try.

[profile_lnd_hangup_0_14_rc2.txt](https://github.com/lightningnetwork/lnd/files/7490490/profile_lnd_hangup_0_14_rc2.txt)

**Comment by bhandras** [comment]
Thanks @xmrk-btc, very grateful for the profile! I'll attempt to put up a PR with the fix.

**Comment by C-Otto** [comment]
I was able to reproduce it, too. Another observation: memory consumption climbed a lot (12.3 GByte VIRT, 2.4 GByte RES).

[goroutine.txt](https://github.com/lightningnetwork/lnd/files/7491196/goroutine.txt)

**Comment by C-Otto** [comment]
Here's the general trend. That's not just lnd, but lnd is the one that grew.

![memory_1d](https://user-images.githubusercontent.com/344948/140622986-92ff182f-8982-41a4-9077-dcab9711a559.png)

**Comment by Roasbeef** [comment]
Makes sense that memory consumption would reise if this happens since a bunch of things build up in internal queues until their saturation point.

**Comment by Roasbeef** [comment]
Should've been fixed by #5941, will re-open if we get a repro that indicates otherwise
